CSE 5243 - Yuxiao Zhao (zhao.2379) & Shuo Lin (lin.2237) - April 3 2018

# Lab 4 — Association Rule Mining

In this lab, you'll preprocess a dataset and use the [MLxtend](http://rasbt.github.io/mlxtend/) machine
learning library to mine frequent patterns and association rules.

Please follow all instructions carefully. There are 16 steps. When you are finished, you
will submit your work via Carmen as usual.

**This lab is due on Tuesday, April 3, 2018, at 11:59 PM ET.**


### BEFORE YOU BEGIN: Install the MLxtend machine learning library.

If you're using anaconda, this is realy easy! At the command prompt, enter the following:

```
$ conda install -c conda-forge mlxtend
```

Answer 'yes' to install any dependencies or updates. That's it! If you run into trouble,
[email me right away](mailto:burkhardt.5@osu.edu)!

In [1]:
# These are the libraries you are most likely to need, but you may include others as well.
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

## Part 1. Download the data.

For this lab, you will use the [Zoo dataset](https://archive.ics.uci.edu/ml/datasets/Zoo), which 
is available for download at the UCI Machine Learning Repository.

### 1. Download the file (`zoo.data`). The file should contain 101 rows $\times$ 18 columns.


In [2]:
# If you don't already have a copy of the Iris data file, you can download it here.
DATAURL = "https://archive.ics.uci.edu/ml/machine-learning-databases/zoo/zoo.data"
import urllib
urllib.request.urlretrieve (DATAURL, "zoo.data")

('zoo.data', <http.client.HTTPMessage at 0x1af1fa5b080>)

## Part 2. Preprocessing

### 2. Use the input data to generate a new data frame with 24 columns, as follows:

| Column No. | Column Name | Description |
|-|-|-|
|1|name| The name of the animal (string)|
|2-16|hair, feathers, ...| All (15) asymmetric binary attributes from the original data file (hair, feathers, etc.)|
|17|has_legs| Does the animal have legs? (1=yes, 0=no) |
|18|mammal|Is the animal a mammal? (1=yes, 0=no) |
|19|bird|Is the animal a bird? (1=yes, 0=no) |
|20|reptile|Is the animal a reptile? (1=yes, 0=no) |
|21|fish|Is the animal a fish? (1=yes, 0=no) |
|22|amphibian|Is the animal an amphibian? (1=yes, 0=no) |
|23|insect|Is the animal an insect? (1=yes, 0=no) |
|24|mollusk|Is the animal a mollusk? (1=yes, 0=no) |

In [3]:
zoo = pd.read_table("zoo.data",sep=',',header = None, names = ["name","hair","feathers","eggs","milk","airborne","aquatic","predator","toothed","backbone","breathes","venomous","fins","legs","tail","domestic","catsize","type","has_legs","mammal","bird","reptile","fish","amphibian","insect","mollusk"])

In [4]:
zoo['has_legs'] = np.where(zoo['legs']>0, 1, 0)
zoo['mammal'] =  np.where(zoo['type']==1, 1, 0)
zoo['bird'] = np.where(zoo['type']==2, 1, 0)
zoo['reptile'] = np.where(zoo['type']==3, 1, 0)
zoo['fish'] = np.where(zoo['type']==4, 1, 0)
zoo['amphibian'] = np.where(zoo['type']==5, 1, 0)
zoo['insect'] = np.where(zoo['type']==6, 1, 0)
zoo['mollusk'] = np.where(zoo['type']==7, 1, 0)

In [5]:
zoo = zoo.drop(['legs','type'],axis=1)
zoo.head()

name  hair  feathers  eggs  milk  airborne  aquatic  predator  toothed  \
0  aardvark     1         0     0     1         0        0         1        1   
1  antelope     1         0     0     1         0        0         0        1   
2      bass     0         0     1     0         0        1         1        1   
3      bear     1         0     0     1         0        0         1        1   
4      boar     1         0     0     1         0        0         1        1   

   backbone   ...     domestic  catsize  has_legs  mammal  bird  reptile  \
0         1   ...            0        1         1       1     0        0   
1         1   ...            0        1         1       1     0        0   
2         1   ...            0        0         0       0     0        0   
3         1   ...            0        1         1       1     0        0   
4         1   ...            0        1         1       1     0        0   

   fish  amphibian  insect  mollusk  
0     0          0       0        0  
1     0          0       0        0  
2     1          0       0        0  
3     0          0       0        0  
4     0          0       0        0  

[5 rows x 24 columns]

### 3. Save the data frame to a CSV file and give it a descriptive name (e.g. `zoo_basket.csv`).

In [6]:
zoo.to_csv("zoo_basket.csv")

## Part 3. Generate Frequent Patterns

### 4. Use the [MLxtend **apriori**](http://rasbt.github.io/mlxtend/api_subpackages/mlxtend.frequent_patterns/#apriori) function to find frequent patterns with $min\_support=0.5$ and $max\_len=5$.

In [7]:
apriori(zoo.iloc[:,1:], min_support=0.5, use_colnames=True, max_len=5)

support                              itemsets
0   0.584158                                [eggs]
1   0.554455                            [predator]
2   0.603960                             [toothed]
3   0.821782                            [backbone]
4   0.792079                            [breathes]
5   0.742574                                [tail]
6   0.772277                            [has_legs]
7   0.603960                   [toothed, backbone]
8   0.514851                       [toothed, tail]
9   0.683168                  [backbone, breathes]
10  0.732673                      [backbone, tail]
11  0.633663                  [backbone, has_legs]
12  0.603960                      [breathes, tail]
13  0.722772                  [breathes, has_legs]
14  0.564356                      [tail, has_legs]
15  0.514851             [toothed, backbone, tail]
16  0.594059            [backbone, breathes, tail]
17  0.633663        [backbone, breathes, has_legs]
18  0.554455            [backbone, tail, has_legs]
19  0.564356            [breathes, tail, has_legs]
20  0.554455  [backbone, breathes, tail, has_legs]

### 5. How many frequent itemsets were generated? List all the frequent 3-itemsets.

  21 frequent itemsets were generated, all the frequent 3-itemsets are the following:

[toothed, backbone, tail], [backbone, breathes, tail], [backbone, breathes, has_legs], [backbone, tail, has_legs], 
[breathes, tail, has_legs]

### 6. Experiment with different values of $min\_support$ and $max\_len$.
* Are you able to find non-zero values for the parameters $min\_support$ and $max\_len$
  that yield frequent itemsets containing all 7 of the animal type items?
* How many frequent patterns were generated using these parameters?

In [8]:
zoo_set = apriori(zoo.iloc[:,1:], min_support=0.03, use_colnames=True, max_len=1)
for i in range(len(zoo_set.iloc[:,1])):
    if 'mammal' in zoo_set.iloc[i,1]:
        print('mammal is in the itemsets')
        break
    if i == len(zoo_set.iloc[:,1])-1:
        print('mammal is not in the itemsets')
        
for i in range(len(zoo_set.iloc[:,1])):
    if 'bird' in zoo_set.iloc[i,1]:
        print('bird is in the itemsets')
        break
    if i == len(zoo_set.iloc[:,1])-1:
        print('bird is not in the itemsets')
        
for i in range(len(zoo_set.iloc[:,1])):
    if 'reptile' in zoo_set.iloc[i,1]:
        print('reptile is in the itemsets')
        break
    if i == len(zoo_set.iloc[:,1])-1:
        print('reptile is not in the itemsets')
        
for i in range(len(zoo_set.iloc[:,1])):
    if 'fish' in zoo_set.iloc[i,1]:
        print('fish is in the itemsets')
        break
    if i == len(zoo_set.iloc[:,1])-1:
        print('fish is not in the itemsets')
        
for i in range(len(zoo_set.iloc[:,1])):
    if 'amphibian' in zoo_set.iloc[i,1]:
        print('amphibian is in the itemsets')
        break
    if i == len(zoo_set.iloc[:,1])-1:
        print('amphibian is not in the itemsets')
        
for i in range(len(zoo_set.iloc[:,1])):
    if 'insect' in zoo_set.iloc[i,1]:
        print('insect is in the itemsets')
        break
    if i == len(zoo_set.iloc[:,1])-1:
        print('insect is not in the itemsets')
        
for i in range(len(zoo_set.iloc[:,1])):
    if 'mollusk' in zoo_set.iloc[i,1]:
        print('mollusk is in the itemsets')
        break
    if i == len(zoo_set.iloc[:,1])-1:
        print('mollusk is not in the itemsets')

mammal is in the itemsets
bird is in the itemsets
reptile is in the itemsets
fish is in the itemsets
amphibian is in the itemsets
insect is in the itemsets
mollusk is in the itemsets


In [9]:
zoo_set.shape

(23, 2)

23 frequent itemsets are generated using the parameters: min_support=0.03, max_len=1

## Part 4. Association Rule Mining

### 7. What is the maximum number of association rules that can be extracted from this dataset, including rules that have zero support?

There are 23 attributes in the datasets. Each attribute could be in the antecedent, consequent, or not the rule (three possibilities). It is not possible that there is one or zero attribute in the rule. 
3^23-2^24+1 = 94,126,401,612 is the maximum number of rules that can be extracted from this dataset.

### 8. Regenerate frequent patterns using $min\_support=0.1$ and $max\_len=5$.

In [10]:
zoo_freqsets = apriori(zoo.iloc[:,1:], min_support=0.1, use_colnames=True, max_len=5)
zoo_freqsets.head()

support    itemsets
0  0.425743      [hair]
1  0.198020  [feathers]
2  0.584158      [eggs]
3  0.405941      [milk]
4  0.237624  [airborne]

### 9. Use the [MLxtend **association_rules**](http://rasbt.github.io/mlxtend/api_subpackages/mlxtend.frequent_patterns/#association_rules) function to find association rules with $min\_threshold=0.7$.

In [11]:
zoo_rules = association_rules(zoo_freqsets, metric='confidence', min_threshold=0.7)
zoo_rules.head()

antecedants consequents  antecedent support  consequent support   support  \
0      (hair)      (milk)            0.425743            0.405941  0.386139   
1      (milk)      (hair)            0.405941            0.425743  0.386139   
2      (hair)   (toothed)            0.425743            0.603960  0.376238   
3      (hair)  (backbone)            0.425743            0.821782  0.386139   
4      (hair)  (breathes)            0.425743            0.792079  0.425743   

   confidence     lift  leverage  conviction  
0    0.906977  2.23426  0.213312    6.386139  
1    0.951220  2.23426  0.213312   11.772277  
2    0.883721  1.46321  0.119106    3.405941  
3    0.906977  1.10367  0.036271    1.915842  
4    1.000000  1.26250  0.088521         inf

### 10. How many association rules were found?

In [12]:
zoo_rules.shape[0]

8584

8584 association rules were found.

### 11. Construct a contingency table for the rule

$$ \{predator, catsize, toothed, mammal\} \rightarrow \{tail\} $$

In [13]:
rule_table = pd.DataFrame(index = ['{predator,catsize,toothed,mammal}','not {predator,catsize,toothed,mammal}','total'],
                         columns = ['tail','not tail', 'total'])

In [14]:
rule_table.loc['{predator,catsize,toothed,mammal}','tail'] = len(zoo[(zoo['predator']==1) & (zoo['catsize']==1) & (zoo['toothed']==1) & (zoo['mammal']==1) & (zoo['tail']==1)])
rule_table.loc['{predator,catsize,toothed,mammal}','not tail'] = len(zoo[(zoo['predator']==1) & (zoo['catsize']==1) & (zoo['toothed']==1) & (zoo['mammal']==1) & (zoo['tail']==0)])
rule_table.loc['{predator,catsize,toothed,mammal}','total'] = len(zoo[(zoo['predator']==1) & (zoo['catsize']==1) & (zoo['toothed']==1) & (zoo['mammal']==1)])
rule_table.loc['not {predator,catsize,toothed,mammal}','tail'] = len(zoo[((zoo['predator']==0) | (zoo['catsize']==0) | (zoo['toothed']==0) | (zoo['mammal']==0)) & (zoo['tail']==1)])
rule_table.loc['not {predator,catsize,toothed,mammal}','not tail'] = len(zoo[((zoo['predator']==0) | (zoo['catsize']==0) | (zoo['toothed']==0) | (zoo['mammal']==0)) & (zoo['tail']==0)])
rule_table.loc['not {predator,catsize,toothed,mammal}','total'] = len(zoo[(zoo['predator']==0) | (zoo['catsize']==0) | (zoo['toothed']==0) | (zoo['mammal']==0)])
rule_table.loc['total','tail'] = len(zoo[(zoo['tail']==1)])
rule_table.loc['total','not tail'] = len(zoo[(zoo['tail']==0)])
rule_table.loc['total','total'] = len(zoo)
rule_table

tail  not tail  total
{predator,catsize,toothed,mammal}        15         4     19
not {predator,catsize,toothed,mammal}    60        22     82
total                                    75        26    101

### 12. Use the contingency table to compute support for the rule. Does your result agree with the output from the *association_rules* function?

In [15]:
support = rule_table.iloc[0,0]/rule_table.iloc[2,2]
print('support of the rule is %f'%support)

support of the rule is 0.148515


In [16]:
zoo_rules.iloc[8077,:]

antecedants           (predator, tail, catsize, toothed)
consequents                                     (mammal)
antecedent support                              0.188119
consequent support                              0.405941
support                                         0.148515
confidence                                      0.789474
lift                                              1.9448
leverage                                       0.0721498
conviction                                       2.82178
Name: 8077, dtype: object

It agrees with the association_rules function.

### 13. List the "most interesting" rules—those with the highest measures of confidence, lift, and leverage. If there are ties (i.e. more than one rule sharing the highest value) then show them all.

In [17]:
# Those with highest leverage
zoo_rules_max = zoo_rules[(zoo_rules['leverage'] == (zoo_rules['leverage'].max()))]
zoo_rules_max

antecedants                   consequents  \
30                        (mammal)                        (milk)   
31                          (milk)                      (mammal)   
405             (backbone, mammal)                        (milk)   
406               (backbone, milk)                      (mammal)   
408                       (mammal)              (backbone, milk)   
409                         (milk)            (backbone, mammal)   
421             (breathes, mammal)                        (milk)   
422               (breathes, milk)                      (mammal)   
424                       (mammal)              (breathes, milk)   
425                         (milk)            (breathes, mammal)   
2305  (backbone, breathes, mammal)                        (milk)   
2306    (backbone, breathes, milk)                      (mammal)   
2309            (backbone, mammal)              (breathes, milk)   
2310              (backbone, milk)            (breathes, mammal)   
2311            (breathes, mammal)              (backbone, milk)   
2312              (breathes, milk)            (backbone, mammal)   
2314                      (mammal)    (backbone, breathes, milk)   
2315                        (milk)  (backbone, breathes, mammal)   

      antecedent support  consequent support   support  confidence      lift  \
30              0.405941            0.405941  0.405941         1.0  2.463415   
31              0.405941            0.405941  0.405941         1.0  2.463415   
405             0.405941            0.405941  0.405941         1.0  2.463415   
406             0.405941            0.405941  0.405941         1.0  2.463415   
408             0.405941            0.405941  0.405941         1.0  2.463415   
409             0.405941            0.405941  0.405941         1.0  2.463415   
421             0.405941            0.405941  0.405941         1.0  2.463415   
422             0.405941            0.405941  0.405941         1.0  2.463415   
424             0.405941            0.405941  0.405941         1.0  2.463415   
425             0.405941            0.405941  0.405941         1.0  2.463415   
2305            0.405941            0.405941  0.405941         1.0  2.463415   
2306            0.405941            0.405941  0.405941         1.0  2.463415   
2309            0.405941            0.405941  0.405941         1.0  2.463415   
2310            0.405941            0.405941  0.405941         1.0  2.463415   
2311            0.405941            0.405941  0.405941         1.0  2.463415   
2312            0.405941            0.405941  0.405941         1.0  2.463415   
2314            0.405941            0.405941  0.405941         1.0  2.463415   
2315            0.405941            0.405941  0.405941         1.0  2.463415   

      leverage  conviction  
30    0.241153         inf  
31    0.241153         inf  
405   0.241153         inf  
406   0.241153         inf  
408   0.241153         inf  
409   0.241153         inf  
421   0.241153         inf  
422   0.241153         inf  
424   0.241153         inf  
425   0.241153         inf  
2305  0.241153         inf  
2306  0.241153         inf  
2309  0.241153         inf  
2310  0.241153         inf  
2311  0.241153         inf  
2312  0.241153         inf  
2314  0.241153         inf  
2315  0.241153         inf

In [18]:
# Those with higheset lift
zoo_rules_max = zoo_rules[(zoo_rules['lift'] == (zoo_rules['lift'].max()))]
zoo_rules_max

antecedants                      consequents  \
344                      (eggs, fins)                           (fish)   
346                            (fish)                     (eggs, fins)   
1880            (eggs, aquatic, fins)                           (fish)   
1883                     (eggs, fins)                  (aquatic, fish)   
1887                  (aquatic, fish)                     (eggs, fins)   
1889                           (fish)            (eggs, aquatic, fins)   
1935            (eggs, fins, toothed)                           (fish)   
1938                     (eggs, fins)                  (fish, toothed)   
1942                  (fish, toothed)                     (eggs, fins)   
1944                           (fish)            (eggs, fins, toothed)   
1973           (backbone, eggs, fins)                           (fish)   
1976                 (backbone, fish)                     (eggs, fins)   
1979                     (eggs, fins)                 (backbone, fish)   
1981                           (fish)           (backbone, eggs, fins)   
2029               (eggs, tail, fins)                           (fish)   
2032                     (eggs, fins)                     (fish, tail)   
2033                     (fish, tail)                     (eggs, fins)   
2036                           (fish)               (eggs, tail, fins)   
6300   (fins, eggs, aquatic, toothed)                           (fish)   
6304            (fins, eggs, toothed)                  (aquatic, fish)   
6307         (aquatic, fish, toothed)                     (eggs, fins)   
6311            (eggs, aquatic, fins)                  (fish, toothed)   
6313                  (fish, toothed)            (eggs, aquatic, fins)   
6316                     (eggs, fins)         (aquatic, fish, toothed)   
6318                  (aquatic, fish)            (fins, eggs, toothed)   
6320                           (fish)   (fins, eggs, aquatic, toothed)   
6364  (backbone, eggs, aquatic, fins)                           (fish)   
6368           (backbone, eggs, fins)                  (aquatic, fish)   
6370        (backbone, aquatic, fish)                     (eggs, fins)   
6374            (eggs, aquatic, fins)                 (backbone, fish)   
...                               ...                              ...   
6409               (eggs, tail, fins)                  (aquatic, fish)   
6412            (aquatic, fish, tail)                     (eggs, fins)   
6415                     (eggs, fins)            (aquatic, fish, tail)   
6417                  (aquatic, fish)               (eggs, tail, fins)   
6418                     (fish, tail)            (eggs, aquatic, fins)   
6421                           (fish)      (eggs, aquatic, tail, fins)   
6460  (backbone, eggs, fins, toothed)                           (fish)   
6464        (backbone, fish, toothed)                     (eggs, fins)   
6467           (backbone, eggs, fins)                  (fish, toothed)   
6470            (fins, eggs, toothed)                 (backbone, fish)   
6473                 (backbone, fish)            (fins, eggs, toothed)   
6475                  (fish, toothed)           (backbone, eggs, fins)   
6478                     (eggs, fins)        (backbone, fish, toothed)   
6480                           (fish)  (backbone, eggs, fins, toothed)   
6501      (fins, eggs, tail, toothed)                           (fish)   
6505            (fins, eggs, toothed)                     (fish, tail)   
6508            (fish, tail, toothed)                     (eggs, fins)   
6512               (eggs, tail, fins)                  (fish, toothed)   
6514                  (fish, toothed)               (eggs, tail, fins)   
6517                     (eggs, fins)            (fish, tail, toothed)   
6519                     (fish, tail)            (fins, eggs, toothed)   
6521                           (fish)      (fins, eggs, tail, toothed)   
6567     (backbone, eggs, tail

In [19]:
# Those with higheset confidence
zoo_rules_max = zoo_rules[(zoo_rules['confidence'] == (zoo_rules['confidence'].max()))]
zoo_rules_max

antecedants  \
4                                    (hair)   
9                                (feathers)   
11                               (feathers)   
12                               (feathers)   
13                               (feathers)   
14                               (feathers)   
15                               (feathers)   
16                                   (bird)   
21                                   (bird)   
22                                   (fish)   
24                                   (milk)   
25                                   (milk)   
30                                 (mammal)   
31                                   (milk)   
33                               (airborne)   
35                               (airborne)   
39                                   (fins)   
41                                   (fish)   
46                                (toothed)   
48                                   (fins)   
52                                   (fish)   
55                                   (fins)   
62                                 (mammal)   
63                                   (bird)   
64                                   (fish)   
71                                 (mammal)   
72                                   (bird)   
74                                   (fish)   
80                                   (bird)   
81                                   (fish)   
...                                     ...   
8462     (catsize, has_legs, tail, toothed)   
8477    (backbone, catsize, has_legs, tail)   
8479    (catsize, has_legs, tail, breathes)   
8484              (catsize, has_legs, tail)   
8493      (backbone, catsize, mammal, tail)   
8495      (catsize, mammal, tail, breathes)   
8501                (catsize, mammal, tail)   
8507     (backbone, mammal, has_legs, tail)   
8509     (breathes, has_legs, mammal, tail)   
8514               (mammal, has_legs, tail)   
8521   (backbone, breathes, bird, has_legs)   
8522       (backbone, bird, has_legs, tail)   
8523       (backbone, breathes, bird, tail)   
8524       (breathes, bird, has_legs, tail)   
8525             (backbone, bird, has_legs)   
8526             (backbone, breathes, bird)   
8527                 (backbone, bird, tail)   
8528             (breathes, bird, has_legs)   
8529                 (bird, has_legs, tail)   
8530                 (breathes, bird, tail)   
8531                       (backbone, bird)   
8532                       (bird, has_legs)   
8533                       (breathes, bird)   
8534                           (bird, tail)   
8535                                 (bird)   
8537  (backbone, catsize, has_legs, mammal)   
8540  (catsize, has_legs, mammal, breathes)   
8547            (catsize, has_legs, mammal)   
8561      (catsize, has_legs, mammal, tail)   
8572      (catsize, has_legs, mammal, tail)   

                               consequents  antecedent support  \
4                               (breathes)            0.425743   
9                                   (eggs)            0.198020   
11                              (backbone)            0.198020   
12                              (breathes)            0.198020   
13                                  (tail)            0.198020   
14                              (has_legs)            0.198020   
15                                  (bird)            0.198020   
16                              (feathers)            0.198020   
21                                  (eggs)            0.198020   
22                                  (eggs)            0.128713   
24                              (backbone)            0.405941   
25                              (breathes)            0.405941   
30                                  (milk)            0.405941   
31                                (mammal)            0.405941   
33                              (breathes)            0.237624   
35                              (has_legs)            0.23762

## Part 5. Association Rules as Predictors

### 14. How many association rules were found in which the consequent contains exactly one element, and that element is one of the 7 animal types?

(Let's call this "Ruleset 2". Save these to another data frame or matrix, and answer
the remaining questions in this section based ONLY ON RULESET2.)


In [20]:
Ruleset2  = zoo_rules[(zoo_rules['consequents'] == {'mammal'}) | (zoo_rules['consequents'] == {'bird'}) | (zoo_rules['consequents'] == {'reptile'}) | 
                      (zoo_rules['consequents'] == {'fish'}) | (zoo_rules['consequents'] == {'amphibian'}) | (zoo_rules['consequents'] == {'insect'}) | (zoo_rules['consequents'] == {'mollusk'})]
Ruleset2.sample(5)

antecedants consequents  antecedent support  \
3318     (predator, hair, milk, breathes)    (mammal)            0.198020   
2271            (has_legs, milk, toothed)    (mammal)            0.366337   
3988     (backbone, has_legs, hair, milk)    (mammal)            0.376238   
8084  (predator, has_legs, tail, toothed)    (mammal)            0.168317   
1314           (backbone, has_legs, hair)    (mammal)            0.376238   

      consequent support   support  confidence      lift  leverage  conviction  
3318            0.405941  0.198020    1.000000  2.463415  0.117636         inf  
2271            0.405941  0.366337    1.000000  2.463415  0.217626         inf  
3988            0.405941  0.376238    1.000000  2.463415  0.223507         inf  
8084            0.405941  0.148515    0.882353  2.173601  0.080188    5.049505  
1314            0.405941  0.376238    1.000000  2.463415  0.223507         inf

In [21]:
Ruleset2.shape[0]

317

There are 317 rules in this case.

### 15. How many rules are there for each animal type?

In [22]:
print("There are %d rules for mammal."%len(Ruleset2[Ruleset2['consequents']=={'mammal'}]))
print("There are %d rules for bird."%len(Ruleset2[Ruleset2['consequents']=={'bird'}]))
print("There are %d rules for reptile."%len(Ruleset2[Ruleset2['consequents']=={'reptile'}]))
print("There are %d rules for fish."%len(Ruleset2[Ruleset2['consequents']=={'fish'}]))
print("There are %d rules for insect."%len(Ruleset2[Ruleset2['consequents']=={'insect'}]))
print("There are %d rules for amphibian."%len(Ruleset2[Ruleset2['consequents']=={'amphibian'}]))
print("There are %d rules for mollusk."%len(Ruleset2[Ruleset2['consequents']=={'mollusk'}]))

There are 215 rules for mammal.
There are 71 rules for bird.
There are 0 rules for reptile.
There are 31 rules for fish.
There are 0 rules for insect.
There are 0 rules for amphibian.
There are 0 rules for mollusk.


## 6. Thought Experiment

Don't implement anything for this section. Just write out your answer in a markdown cell.

### 16. Explain how you might use association rules to suggest an animal's type based on the other attributes. Based on your observations, what changes would you make in the frequent itemset generation and rule generation steps?

I will select the rules with highest support and confidence which has the one of the animal type as the consequent. Use those antecdents to predict the animal type.

I will increase the max_length, decrease the confidence threshold or decrease min_support to regenerate frequent itemsets and associate rules which will cover at least six of all the animal types. For each animal type, I will choose one rule with highest support and confidence. Last I will use these rules to predict the animal type using these attributes.

## 7. Submit your work

**Submit your completed notebook, along with the data file you created in Part 2, via Carmen as LAB 4.**